lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart

```diff
@@ -0,0 +1,690 @@
+    for (_UnicodeRange range in subset.ranges) {
+      _resolvedNotoTreeRoot =
+          _insertNotoFontRange(range, subset, _resolvedNotoTreeRoot);
+    }
+  }
+}
+
+/// Finds the minimum set of fonts which covers all of the [codeunits].
+///
+/// Since set cover is NP-complete, we approximate using a greedy algorithm
+/// which finds the font which covers the most codeunits. If multiple CJK
+/// fonts match the same number of codeunits, we choose one based on the user's
+/// locale.
+Set<_NotoFont> _findMinimumFontsForCodeunits(
+    List<int> codeunits, Set<_NotoFont> fonts) {
+  List<int> unmatchedCodeunits = List<int>.from(codeunits);
+  Set<_NotoFont> minimumFonts = <_NotoFont>{};
+  List<_NotoFont> bestFonts = <_NotoFont>[];
+  int maxCodeunitsCovered = 0;
+
+  while (unmatchedCodeunits.isNotEmpty) {
+    for (var font in fonts) {
+      int codeunitsCovered = 0;
+      for (int codeunit in unmatchedCodeunits) {
+        if (font.matchesCodeunit(codeunit)) {
+          codeunitsCovered++;
+        }
+      }
+      if (codeunitsCovered > maxCodeunitsCovered) {
+        bestFonts.clear();
+        bestFonts.add(font);
+        maxCodeunitsCovered = codeunitsCovered;
+      } else if (codeunitsCovered == maxCodeunitsCovered) {
+        bestFonts.add(font);
+      }
+    }
+    assert(bestFonts.isNotEmpty);
+    // If the list of best fonts are all CJK fonts, choose the best one based
+    // on locale. Otherwise just choose the first font.
+    _NotoFont bestFont = bestFonts.first;
+    if (bestFonts.length > 1) {
+      if (bestFonts.every((font) => _cjkFonts.contains(font))) {
+        String language = html.window.navigator.language;
+        if (language == 'zh-Hans' ||
+            language == 'zh-CN' ||
+            language == 'zh-SG' ||
+            language == 'zh-MY') {
+          if (bestFonts.contains(_notoSansSC)) {
+            bestFont = _notoSansSC;
+          }
+        } else if (language == 'zh-Hant' ||
+            language == 'zh-TW' ||
+            language == 'zh-MO') {
+          if (bestFonts.contains(_notoSansTC)) {
+            bestFont = _notoSansTC;
+          }
+        } else if (language == 'zh-HK') {
+          if (bestFonts.contains(_notoSansHK)) {
+            bestFont = _notoSansHK;
+          }
+        } else if (language == 'ja') {
+          if (bestFonts.contains(_notoSansJP)) {
+            bestFont = _notoSansJP;
+          }
+        }
+      }
+    }
+    unmatchedCodeunits
+        .removeWhere((codeunit) => bestFont.matchesCodeunit(codeunit));
+    minimumFonts.add(bestFont);
+  }
+  return minimumFonts;
+}
+
+void _ensureNotoFontTreeCreated() {
+  if (_notoTreeRoot != null) {
+    return;
+  }
+
+  for (_NotoFont font in _notoFonts) {
+    for (_UnicodeRange range in font.unicodeRanges) {
+      _notoTreeRoot = _insertNotoFontRange(range, font, _notoTreeRoot);
+    }
+  }
+}
+
+List<_NotoFont> _lookupNotoFontsForCodeunit(int codeunit) {
```

Comment:
Do you need this to be List ? Iterable<_NotoFont> may prevent having to convert node.fonts to list.

Reply:
Stopped copying the list